rootless-install.sh

```diff
@@ -0,0 +1,268 @@
+		>&2 echo "Could not find records for the current user $(id -un) from /etc/subuid . Please make sure valid subuid range is set there."
+		exit 1
+	fi
+	if ! grep "$(id -gn)" /etc/subgid >/dev/null 2>&1; then
+		>&2 echo "Could not find records for the current user $(id -un) from /etc/subgid . Please make sure valid subuid range is set there."
+		exit 1
+	fi
+}
+
+start_docker() {
+	tmpdir=$(mktemp -d)
+	mkdir -p $tmpdir/lower $tmpdir/upper $tmpdir/work $tmpdir/merged
+	if "$BIN"/rootlesskit mount -t overlay overlay -olowerdir=$tmpdir/lower,upperdir=$tmpdir/upper,workdir=$tmpdir/work $tmpdir/merged >/dev/null 2>&1; then
+		USE_OVERLAY=1
+	fi
+	rm -rf "$tmpdir"
+
+
+	if [ -z "$SYSTEMD" ]; then
+		start_docker_nonsystemd
+		return
+	fi
+	
+	mkdir -p $HOME/.config/systemd/user
+	
+	DOCKERD_FLAGS="--experimental"
+	
+	if ! which iptables >/dev/null 2>&1 ; then
+		DOCKERD_FLAGS="$DOCKERD_FLAGS --iptables=false"
+	fi
+	
+	if [ "$USE_OVERLAY" = "1" ]; then
+		DOCKERD_FLAGS="$DOCKERD_FLAGS --storage-driver=overlay"
+	else
+		DOCKERD_FLAGS="$DOCKERD_FLAGS --storage-driver=vfs"
+	fi
+	
+	
+	if [ ! -f $HOME/.config/systemd/user/docker.service ]; then
+		cat <<EOT > $HOME/.config/systemd/user/docker.service
+[Unit]
+Description=Docker Application Container Engine (Rootless)
+Documentation=https://docs.docker.com
+
+[Service]
+Environment=PATH=$HOME/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+ExecStart=$HOME/bin/dockerd-rootless.sh $DOCKERD_FLAGS
+ExecReload=/bin/kill -s HUP \$MAINPID
+TimeoutSec=0
+RestartSec=2
+Restart=always
+StartLimitBurst=3
+StartLimitInterval=60s
+LimitNOFILE=infinity
+LimitNPROC=infinity
+LimitCORE=infinity
+TasksMax=infinity
+Delegate=yes
+Type=simple
+
+[Install]
+WantedBy=default.target
+EOT
+	systemctl --user daemon-reload
+	fi
+	if ! systemctl --user status docker >/dev/null 2>&1; then
+		echo "# starting systemd service"
+		systemctl --user start docker
+	fi
+	systemctl --user status docker | cat
+}
+
+service_instructions() {
+	if [ -z "$SYSTEMD" ]; then
+		return
+	fi
+	cat <<EOT
+#
+# To control docker service run:
+# systemctl --user (start|stop|restart) docker
+#
+EOT
+}
+
+
+start_docker_nonsystemd() {
+	echo <<EOT
+# systemd not detected, dockerd daemon needs to be started manually
+#
+$BIN/dockerd-rootless.sh --experimental --iptables=false --storage-driver $driver
+#
+EOT
+}
+
+print_instructions() {
+	start_docker
+	echo "# Docker binaries are installed in $BIN"
+	if [ "$(which $DAEMON)" != "$BIN/$DAEMON" ]; then
+		echo "# WARN: dockerd is not in your current PATH or pointing to $BIN/$DAEMON"
+	fi
+	echo "# Please make sure following environment variables are set (or put them to ~/.bashrc):\n"
+	
+	case :$PATH: in
+	*:$BIN:*) ;;
+	*) echo "export PATH=$BIN:\$PATH" ;;
+	esac
+
+	echo "export DOCKER_HOST=unix://$XDG_RUNTIME_DIR/docker.sock"
+	echo
+	service_instructions
+	exit 0
+}
+
+do_install() {
+	init_vars
+	checks
+
+	set -e
+	tmp=$(mktemp -d)
+	trap "rm -rf $tmp" EXIT INT TERM
```

Comment:
should we remove `trap` ? The other tmpdir is not trapped.

Reply:
I don't mind this as this tmpdir is used for a much longer time (meaning more chances for cancel)